### PR TITLE
Add sparse compressed meta tensor support

### PR DESCRIPTION
**Pull request by pearu**
As in the title.

Replaces https://github.com/pytorch/pytorch/pull/120498 and https://github.com/pytorch/pytorch/pull/120562 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120920
* __->__ #120707
* #120703

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @jcaip

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -3040,12 +3041,17 @@ def hint(s):
...
-                for i, ss in enumerate(curr_t.size()):
-                    property_source = TensorPropertySource(src, TensorProperty.SIZE, i)
-                    track_symint(property_source, ss, constraint[i])
-                for i, ss in enumerate(curr_t.stride()):
-                    track_symint(TensorPropertySource(src, TensorProperty.STRIDE, i), ss)
-                track_symint(TensorPropertySource(src, TensorProperty.STORAGE_OFFSET), curr_t.storage_offset())
+                if is_sparse_any(curr_t):
+                    for i, ss in enumerate(curr_t.size()):
+                        property_source = TensorPropertySource(src, TensorProperty.SIZE, i)
+                        track_symint(property_source, ss, constraint[i])
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:3047]: I feel there are some integers we have to track, like nnz?

**Review by ezyang — APPROVED** [review]
jolly good

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (pearu) [reply torch/fx/experimental/symbolic_shapes.py:3037]: What is the exact function of tracking sparse tensor shape values and other integer parameters? Namely, a sparse tensor can be viewed as a tuple of indices and values tensors. While the sparse tensor shape and nnz is related to the shapes of indices and values, should we also track the strides of indices and values tensors (we do support non-contiguous values, for instance)?

**Comment by laithsakka** [comment]
with this change shall i not graph break on    torch.sparse_compressed_tensor? in 
https://github.com/pytorch/pytorch/pull/120458/ 
@pearu

**Comment by pearu** [comment]
> with this change shall i not graph break on torch.sparse_compressed_tensor? in #120458

If it is not necessary, then no. Note that adding sparse compressed tensors support to torch compiler is WIP, see e.g. https://github.com/pytorch/pytorch/pull/117907

**Pull request closed without merging**